File: src/features/desktop-monitor/DisplayMonitor.ts
import { Shell } from '../../utility/Shell';
import { Logger } from '../../utility/Logger';
import type { MonitorInfo } from './types';

export class DisplayMonitor {
	private static instance: DisplayMonitor;
	private logger = Logger.getInstance();

	private constructor() {}

	static getInstance(): DisplayMonitor {
		if (!DisplayMonitor.instance) {
			DisplayMonitor.instance = new DisplayMonitor();
		}
		return DisplayMonitor.instance;
	}

	isAvailable(): boolean {
		return !!process.env.HYPRLAND_INSTANCE_SIGNATURE;
	}

	async getMonitors(): Promise<MonitorInfo[]> {
		if (!this.isAvailable()) {
			return [];
		}

		try {
			const result = await Shell.execute('hyprctl monitors -j', {
				timeout: 5000,
			});

			if (result.exitCode !== 0) {
				this.logger.error(`hyprctl monitors failed: ${result.stderr}`);
				return [];
			}

			return JSON.parse(result.stdout);
		} catch (error) {
			this.logger.error(`Error getting monitors: ${error}`);
			return [];
		}
	}

	async getPrimaryMonitor(): Promise<MonitorInfo | null> {
		const monitors = await this.getMonitors();
		if (monitors.length === 0) {
			return null;
		}

		const activeMonitor = monitors.find((m) => m.activeWorkspace.id > 0);
		return activeMonitor || monitors[0];
	}

	async getMonitorCount(): Promise<number> {
		const monitors = await this.getMonitors();
		return monitors.length;
	}

	formatMonitorInfo(monitors: MonitorInfo[]): string {
		if (monitors.length === 0) {
			return 'Display Information:\n  No monitors detected';
		}

		const lines = ['Display Information:'];

		for (const monitor of monitors) {
			lines.push('');
			lines.push(`  ${monitor.name}:`);

			if (monitor.description && monitor.description !== monitor.name) {
				lines.push(`    Description: ${monitor.description}`);
			}

			if (monitor.make && monitor.model) {
				lines.push(`    Make/Model: ${monitor.make} ${monitor.model}`);
			}

			lines.push(`    Resolution: ${monitor.width}x${monitor.height}@${monitor.refreshRate.toFixed(2)}Hz`);
			lines.push(`    Position: ${monitor.x},${monitor.y}`);
			lines.push(`    Scale: ${monitor.scale.toFixed(2)}`);
			lines.push(`    VRR: ${monitor.vrr ? 'enabled' : 'disabled'}`);
			lines.push(`    DPMS: ${monitor.dpmsStatus ? 'on' : 'off'}`);
			lines.push(`    Active Workspace: ${monitor.activeWorkspace.name}`);

			if (monitor.transform !== 0) {
				lines.push(`    Transform: ${monitor.transform}`);
			}
		}

		return lines.join('\n');
	}

	formatMonitorSummary(monitors: MonitorInfo[]): string {
		if (monitors.length === 0) {
			return 'No monitors';
		}

		const summaries = monitors.map((m) => {
			return `${m.name} (${m.width}x${m.height}@${m.refreshRate.toFixed(0)}Hz${m.vrr ? ' VRR' : ''})`;
		});

		return summaries.join(', ');
	}
}

File: src/features/desktop-monitor/DisplayMonitor.go
/**
 * Display monitor - monitors display/monitor information
 */

package desktopmonitor

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/ln64-git/daemira/src/utility"
)

// DisplayMonitor monitors display/monitor information
type DisplayMonitor struct {
	logger *utility.Logger
	shell  *utility.Shell
	mu     sync.RWMutex
}

var (
	displayMonitorInstance *DisplayMonitor
	displayMonitorOnce     sync.Once
)

// GetDisplayMonitor returns the singleton DisplayMonitor instance
func GetDisplayMonitor() *DisplayMonitor {
	displayMonitorOnce.Do(func() {
		displayMonitorInstance = &DisplayMonitor{
			logger: utility.GetLogger(),
			shell:  utility.NewShell(utility.GetLogger()),
		}
	})
	return displayMonitorInstance
}

// IsAvailable checks if Hyprland is available
func (dm *DisplayMonitor) IsAvailable() bool {
	return os.Getenv("HYPRLAND_INSTANCE_SIGNATURE") != ""
}

// GetMonitors gets all monitors
func (dm *DisplayMonitor) GetMonitors(ctx context.Context) ([]MonitorInfo, error) {
	if !dm.IsAvailable() {
		return []MonitorInfo{}, nil
	}

	result, err := dm.shell.Execute(ctx, "hyprctl monitors -j", &utility.ExecOptions{
		Timeout: 5 * time.Second,
	})

	if err != nil || result.ExitCode != 0 {
		dm.logger.Error("hyprctl monitors failed: %v", err)
		return []MonitorInfo{}, err
	}

	var monitors []MonitorInfo
	if err := json.Unmarshal([]byte(result.Stdout), &monitors); err != nil {
		dm.logger.Error("Error parsing monitors JSON: %v", err)
		return []MonitorInfo{}, err
	}

	return monitors, nil
}

// GetPrimaryMonitor gets the primary/active monitor
func (dm *DisplayMonitor) GetPrimaryMonitor(ctx context.Context) (*MonitorInfo, error) {
	monitors, err := dm.GetMonitors(ctx)
	if err != nil {
		return nil, err
	}

	if len(monitors) == 0 {
		return nil, nil
	}

	// Find active monitor (with active workspace ID > 0)
	for i := range monitors {
		if monitors[i].ActiveWorkspace.ID > 0 {
			return &monitors[i], nil
		}
	}

	// Return first monitor if no active found
	return &monitors[0], nil
}

// GetMonitorCount gets the number of monitors
func (dm *DisplayMonitor) GetMonitorCount(ctx context.Context) (int, error) {
	monitors, err := dm.GetMonitors(ctx)
	if err != nil {
		return 0, err
	}
	return len(monitors), nil
}

// FormatMonitorInfo formats monitor info for display
func (dm *DisplayMonitor) FormatMonitorInfo(monitors []MonitorInfo) string {
	if len(monitors) == 0 {
		return "Display Information:\n  No monitors detected"
	}

	lines := []string{"Display Information:"}

	for _, monitor := range monitors {
		lines = append(lines, "")
		lines = append(lines, fmt.Sprintf("  %s:", monitor.Name))

		if monitor.Description != "" && monitor.Description != monitor.Name {
			lines = append(lines, fmt.Sprintf("    Description: %s", monitor.Description))
		}

		if monitor.Make != "" && monitor.Model != "" {
			lines = append(lines, fmt.Sprintf("    Make/Model: %s %s", monitor.Make, monitor.Model))
		}

		lines = append(lines, fmt.Sprintf("    Resolution: %dx%d@%.2fHz", monitor.Width, monitor.Height, monitor.RefreshRate))
		lines = append(lines, fmt.Sprintf("    Position: %d,%d", monitor.X, monitor.Y))
		lines = append(lines, fmt.Sprintf("    Scale: %.2f", monitor.Scale))
		lines = append(lines, fmt.Sprintf("    VRR: %s", boolToEnabled(monitor.VRR)))
		lines = append(lines, fmt.Sprintf("    DPMS: %s", boolToOnOff(monitor.DPMSStatus)))
		lines = append(lines, fmt.Sprintf("    Active Workspace: %s", monitor.ActiveWorkspace.Name))

		if monitor.Transform != 0 {
			lines = append(lines, fmt.Sprintf("    Transform: %d", monitor.Transform))
		}
	}

	return strings.Join(lines, "\n")
}

// FormatMonitorSummary formats a summary of monitors
func (dm *DisplayMonitor) FormatMonitorSummary(monitors []MonitorInfo) string {
	if len(monitors) == 0 {
		return "No monitors"
	}

	var summaries []string
	for _, m := range monitors {
		vrrStr := ""
		if m.VRR {
			vrrStr = " VRR"
		}
		summaries = append(summaries, fmt.Sprintf("%s (%dx%d@%.0fHz%s)", m.Name, m.Width, m.Height, m.RefreshRate, vrrStr))
	}

	return strings.Join(summaries, ", ")
}

// boolToEnabled converts bool to "enabled"/"disabled"
func boolToEnabled(b bool) string {
	if b {
		return "enabled"
	}
	return "disabled"
}

// boolToOnOff converts bool to "on"/"off"
func boolToOnOff(b bool) string {
	if b {
		return "on"
	}
	return "off"
}
